### PR TITLE
[AIRFLOW-6251] add config for max tasks per dag

### PR DESCRIPTION
**Pull request by tooptoop4**
---
Issue link: [AIRFLOW-6251](https://issues.apache.org/jira/browse/AIRFLOW-6251)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/models/dagbag.py [hunk 1]
```diff
@@ -300,26 +302,37 @@ def process_file(self, filepath, only_if_updated=True, safe_mode=True):
...
-                            croniter(dag._schedule_interval)
-                        found_dags.append(dag)
-                        found_dags += dag.subdags
-                    except (CroniterBadCronError,
-                            CroniterBadDateError,
-                            CroniterNotAlphaError) as cron_e:
-                        self.log.exception("Failed to bag_dag: %s", dag.full_filepath)
+
+                    num_tasks = len(dag.tasks)
+                    if max_tasks_per_dag > 0 and num_tasks > max_tasks_per_dag:
```
Comment (zhongjiajie) [inline airflow/models/dagbag.py:307]:
```suggestion
                    if 0 < max_tasks_per_dag < num_tasks:
```

**Review by zhongjiajie — COMMENTED** [review]
If want to upload huge dag, should we set default value to not  unlimit? BTW, should we add some doc to describe it?

File: airflow/models/dagbag.py (hunk 1 above)
Comment (ashb) [reply airflow/models/dagbag.py:307]: That doens't wotk @zhongjiajie 

```
In [1]: x = 0                                                                                                                                                  

In [2]: num_task = 10                                                                                                                                          

In [3]: 0 < x < num_task                                                                                                                                       
Out[3]: False
```

Comment (zhongjiajie) [reply airflow/models/dagbag.py:307]: Really? but I remember we have a PR to do that, I will take a look

Comment (zhongjiajie) [reply airflow/models/dagbag.py:307]: @ashb  After some quick look, in
https://github.com/apache/airflow/blob/4a344f13d26ecbb627bb9968895b290bfd86e4da/airflow/cli/commands/webserver_command.py#L90
And it work in 
```py
In [1]: x = 0

In [2]: num_task = 10

In [3]: 0 < x < num_task
Out[3]: False

In [4]: 0 <= x < num_task
Out[4]: True
```

**Comment by tooptoop4** [comment]
> If want to upload huge dag, should we set default value to not unlimit? BTW, should we add some doc to describe it?

existing behaviour is unlimited, i want this to not break existing users, so it is 'opt-in'

**Comment by zhongjiajie** [comment]
> > If want to upload huge dag, should we set default value to not unlimit? BTW, should we add some doc to describe it?
> 
> existing behaviour is unlimited, i want this to not break existing users, so it is 'opt-in'

I prefer to add use fix value for it and change the `UPDATING.md`
If we decide to use unlimited, at least we should add some document to remind users.

File: airflow/models/dagbag.py (hunk 1 above)
Comment (tooptoop4) [reply airflow/models/dagbag.py:307]: confusion and need to test it proves the shorthand syntax is not clear, readable code

**Comment by tooptoop4** [comment]
@zhongjiajie any idea on why test not working?

File: airflow/models/dagbag.py (hunk 1 above)
Comment (zhongjiajie) [reply airflow/models/dagbag.py:307]: But code base already have some example and IDE like pycharm will give a hint change to `var1 < var2 < var3`

File: tests/models/test_dagbag.py [hunk 2]
```diff
@@ -151,6 +151,42 @@ def test_zip(self):
         dagbag.process_file(os.path.join(TEST_DAGS_FOLDER, "test_zip.zip"))
         self.assertTrue(dagbag.get_dag("test_zip_dag"))
 
+    @conf_vars({('core', 'max_tasks_per_dag'): '5'})
+    def test_process_file_max_task_check(self):
+        """
+        test if num_tasks > max_tasks_per_dag can be identified
+        """
+        a_dag_id = "example_short_circuit_operator"
```
Comment (zhongjiajie) [inline tests/models/test_dagbag.py]:
```suggestion
        a_dag_id = "test_example_bash_operator"
```

I take a quick look. `TEST_DAGS_FOLDER` point to `airflow/tests/models/../dags` and this path not including dag `example_short_circuit_operator`. So you could not process_file `airflow/tests/models/../dags/example_short_circuit_operator.py`

Maybe you should use others dag in `airflow/tests/models/../dags`(which `TEST_DAGS_FOLDER` point to), just like `test_example_bash_operator` and change `@conf_vars({('core', 'max_tasks_per_dag'): '5'})` to `@conf_vars({('core', 'max_tasks_per_dag'): '7'})`.

So as others test @tooptoop4

**Comment by zhongjiajie** [comment]
FYI, Travis-CI still sad, this time is timeout, may maybe you should test int you local breeze using `pytest -vs tests/cli/commands/test_task_command.py` for more detail

**Comment by tooptoop4** [comment]
travis happy

**Comment by xinbinhuang** [comment]
I have a probably unwelcome comment. But do we really need to have this config? I feel like this is something that should be checked during CI/CD before it deploys to Airflow. It gives an extra overhead (though trivial) when parsing the file. Also, it makes a bit harder to read the code (which is already a bit hard to read)

**Comment by tooptoop4** [comment]
@xinbinhuang a dynamic dag may generate tasks at runtime based off a list of files in a 3rd party server (ie an sftp drive to be downloaded from and its csvs ingested) so no way for cicd to trap it

**Comment by xinbinhuang** [comment]
You can check the task count in the CI/CD process after deploying to source control similar to other DAG validation tests that you may have ([relevant blog post](https://blog.usejournal.com/testing-in-airflow-part-1-dag-validation-tests-dag-definition-tests-and-unit-tests-2aa94970570c) & [gist](https://gist.github.com/criccomini/2862667822af7fae8b55682faef029a7). A simple example:

```python
import unittest

from airflow.models import DagBag

class TestTaskCountPerDag(unittest.TestCase):
    """
    Test the if the number of tasks per dag exceed the threshold
    """

    MAX_TASKS_PER_DAG = 3

    def _get_dagbag(self):
        dag_folder = os.getenv('AIRFLOW_DAGS', False)
        self.assertTrue(
            dag_folder,
            'AIRFLOW_DAGS must be set to a folder that has DAGs in it.')
        return DagBag(dag_folder=dag_folder, include_examples=False)

    def test_max_tasks_per_dag(self):
        """
        Verify that tasks per dag do not exceed the threshold
        """
        dagbag = self._get_dagbag()
        dags = dagbag.dags

        for dag_name, dag in dags.items():
            self.assertLessEqual(dag.task_count, 
                                 MAX_TASKS_PER_DAG,
                                 msg=('Numer of tasks for DAG {} is greater than ' + 
                                     'the threshold {}').format(dag_name, MAX_TASKS_PER_DAG)
```

**Comment by tooptoop4** [comment]
@kaxil WDYT

**Comment by kaxil** [comment]
> @kaxil WDYT

I agree with @xinbinhuang that we might not need this setting after all ! Although that being said, create a post on dev mailing list and see if there is any interest in this feature, if so we can think about it.

**Comment by tooptoop4** [comment]
be gone stalebot

**Comment by mik-laj** [comment]
Why not add this code as an example for a cluster policy? I don't think it has to be a separate option. https://airflow.readthedocs.io/en/latest/concepts.html#cluster-policy

**Review by mik-laj — CHANGES_REQUESTED** [review]
Why not add this code as an example for a cluster policy? I don't think it has to be a separate option. https://airflow.readthedocs.io/en/latest/concepts.html#cluster-policy

**Comment by tooptoop4** [comment]
crispy

**Pull request closed without merging**